### PR TITLE
Simplification rules for bool2Word

### PR DESCRIPTION
**Pull request by ttuegel**
The Haskell backend will use these simplification rules to simplify the common
patterns `bool2Word(B) ==Int 0` and `bool2Word(B) ==Int 1` to fix a performance
regression. Previously, the frontend emitted rules that generalized `==Int` to
`==K`. Then, the backend would branch on `==K`. However, branching causes other
performance problems, so the simplification rules seem like a good compromise.

Fixes kframework/kore#1686

### CODE REVIEW

File: evm-types.md [hunk 1]
```diff
@@ -138,6 +138,9 @@ Primitives provide the basic conversion from K's sorts `Int` and `Bool` to EVM's
     rule bool2Word( B:Bool ) => 1 requires B
     rule bool2Word( B:Bool ) => 0 requires notBool B
 
+    rule bool2Word( B:Bool ) ==Int 1 => B         [simplification]
+    rule bool2Word( B:Bool ) ==Int 0 => notBool B [simplification]
+
```
Comment (ehildenb) [inline evm-types.md]: My guess is that these need to go in `tests/specs/lemmas.k`, in the appropriate module where only the Haskell backend will see them @ttuegel .

Comment (ttuegel) [reply evm-types.md]: Done!

**Comment by ttuegel** [comment]
Although this brings performance to an acceptable level, I am also working on a change in the frontend which will improve performance further by keeping term size smaller.

**Comment by ehildenb** [comment]
Awesome, this looks good :)

**Comment by ttuegel** [comment]
It seems to still time-out on some proof.

**Comment by ehildenb** [comment]
I profiled it locally (but after updating to latest KEVM), and it doesn't seem to increase overall runtime at all, so I'm surprised. I just merged with master and maybe it has an effect.

Note that also the frontend parsing ones (dry-run with haskell backend) are timing out, so maybe it's a frontend issue addressed by the last update to the frontend that happened.

**Comment by ehildenb** [comment]
These are the times I have:

```
*   4a9950960 - Merge remote-tracking branch 'upstream/master' into testing                      - Everett Hildenbrandt (HEAD, testing)
|\  testset: prove - Fri Mar 20 11:10:00 UTC 2020
| | 0  0.29s     0.01s      0.20s     3112KB     rm    -rf         deps           .build
| | 0  1.20s     0.71s      0.38s     30744KB    git   submodule   update         --init  --recursive
| | 0  214.89s   1622.81s   2088.39s  2839820KB  make  deps        RELEASE=true
| | 0  47.02s    1385.14s   31.54s    2760184KB  make  build-java  build-haskell  -j8
| | 0  1363.99s  43810.20s  2980.59s  4803952KB  make  test-prove  -j8            -k      TEST='profile  .build/logs/prove  timeout  400  ./kevm'  ||  true
| | 
| * 13d03fd07 - Use pipes instead of a socket for communication with plugin (#755)               - Dwight Guth (upstream/master)
| | testset: prove - Fri Mar 20 10:42:51 UTC 2020
| | 0  0.30s     0.01s      0.29s     3256KB     rm    -rf         deps           .build
| | 0  1.03s     0.72s      0.35s     31048KB    git   submodule   update         --init  --recursive
| | 0  218.74s   1909.26s   1986.23s  2778136KB  make  deps        RELEASE=true
| | 0  47.20s    1387.77s   32.17s    2805520KB  make  build-java  build-haskell  -j8
| | 0  1359.21s  41698.22s  2084.76s  4833436KB  make  test-prove  -j8            -k      TEST='profile  .build/logs/prove  timeout  400  ./kevm'  ||  true
```

**Pull request merged**